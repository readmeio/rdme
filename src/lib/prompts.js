const semver = require('semver');
const { prompt } = require('enquirer');
const parse = require('parse-link-header');

exports.generatePrompts = (versionList, selectOnly = false) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to use an existing project version or create a new one?',
    skip() {
      return selectOnly;
    },
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return selectOnly ? false : this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return selectOnly ? true : this.enquirer.answers.option === 'update';
    },
    hint: '1.0.0',
  },
];

function specOptions(specList, parsedDocs, currPage, totalPages) {
  const specs = specList.map(s => {
    return {
      message: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs.prev.page) specs.push({ message: `< Prev (page ${currPage - 1} of ${totalPages})`, value: 'prev' });
  if (parsedDocs.next.page) {
    specs.push({ message: `Next (page ${currPage + 1} of ${totalPages}) >`, value: 'next' });
  }
  return specs;
}

const updateOasPrompt = (specList, parsedDocs, currPage, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async result(spec) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await newSpecs.json();
          const { specId } = await prompt(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs)
          );
          return specId;
        } catch (e) {
          return null;
        }
      }
      return spec;
    },
  },
];

exports.createOasPrompt = (specList, parsedDocs, totalPages, getSpecs) => [
  {
    type: 'select',
    name: 'option',
    message: 'Would you like to update an existing OAS file or create a new one?',
    choices: [
      { message: 'Update existing', value: 'update' },
      { message: 'Create a new spec', value: 'create' },
    ],
    async result(picked) {
      if (picked === 'update') {
        try {
          const { specId } = await prompt(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        } catch (e) {
          return null;
        }
      }
      return picked;
    },
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    initial: opts.newVersion || false,
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
    validate(val) {
      return semver.valid(semver.coerce(val)) ? true : this.styles.danger('Please specify a semantic version.');
    },
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || isUpdate?.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
