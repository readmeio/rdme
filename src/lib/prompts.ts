import type { Version } from '../cmds/versions';
import type { Response } from 'node-fetch';
import type { Choice, PromptObject } from 'prompts';

import parse from 'parse-link-header';
import semver from 'semver';

import promptTerminal from './promptWrapper';
import { handleRes } from './readmeAPIFetch';

interface Spec {
  _id: string;
  title: string;
}

type SpecList = Spec[];

interface ParsedDocs {
  next?: {
    page: number;
    url: string;
  };
  prev?: {
    page: number;
    url: string;
  };
}

function specOptions(specList: SpecList, parsedDocs: ParsedDocs, currPage: number, totalPages: number): Choice[] {
  const specs = specList.map(s => {
    return {
      description: `API Definition ID: ${s._id}`, // eslint-disable-line no-underscore-dangle
      title: s.title,
      value: s._id, // eslint-disable-line no-underscore-dangle
    };
  });
  if (parsedDocs?.prev?.page) {
    specs.push({
      description: 'Go to the previous page',
      title: `< Prev (page ${currPage - 1} of ${totalPages})`,
      value: 'prev',
    });
  }
  if (parsedDocs?.next?.page) {
    specs.push({
      description: 'Go to the next page',
      title: `Next (page ${currPage + 1} of ${totalPages}) >`,
      value: 'next',
    });
  }
  return specs;
}

const updateOasPrompt = (
  specList: SpecList,
  parsedDocs: ParsedDocs,
  currPage: number,
  totalPages: number,
  getSpecs: (url: string) => Promise<Response>,
): PromptObject[] => [
  {
    type: 'select',
    name: 'specId',
    message: 'Select your desired file to update',
    choices: specOptions(specList, parsedDocs, currPage, totalPages),
    async format(spec: string) {
      if (spec === 'prev') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.prev.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage - 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      } else if (spec === 'next') {
        try {
          const newSpecs = await getSpecs(`${parsedDocs.next.url}`);
          const newParsedDocs = parse(newSpecs.headers.get('link'));
          const newSpecList = await handleRes(newSpecs);
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(
            updateOasPrompt(newSpecList, newParsedDocs, currPage + 1, totalPages, getSpecs),
          );
          return specId;
        } catch (e) {
          return null;
        }
      }

      return spec;
    },
  },
];

export function createOasPrompt(
  specList: SpecList,
  parsedDocs: ParsedDocs,
  totalPages: number,
  getSpecs: ((url: string) => Promise<Response>) | null,
): PromptObject[] {
  return [
    {
      type: 'select',
      name: 'option',
      message: 'Would you like to update an existing OAS file or create a new one?',
      choices: [
        { title: 'Update existing', value: 'update' },
        { title: 'Create a new spec', value: 'create' },
      ],
      async format(picked: 'update' | 'create') {
        if (picked === 'update') {
          // @todo: figure out how to add a stricter type here, see:
          // https://github.com/readmeio/rdme/pull/570#discussion_r949715913
          const { specId } = await promptTerminal(updateOasPrompt(specList, parsedDocs, 1, totalPages, getSpecs));
          return specId;
        }

        return picked;
      },
    },
  ];
}

/**
 * Series of prompts to construct a version object,
 * used in our `versions:create` and `versions:update` commands
 */
export function versionPrompt(
  /** list of versions, used for prompt about which version to fork */
  versionList: Version[],
  /** existing version if we're performing an update */
  isUpdate?: {
    is_stable: boolean;
  },
): PromptObject[] {
  return [
    {
      // only runs for versions:create command
      type: isUpdate ? null : 'select',
      name: 'from',
      message: 'Which version would you like to fork from?',
      choices: versionList.map(v => {
        return {
          title: v.version,
          value: v.version,
        };
      }),
    },
    {
      // only runs for versions:update command
      type: !isUpdate ? null : 'text',
      name: 'newVersion',
      message: 'What should the version be renamed to?',
      hint: '1.0.0',
      validate(val: string) {
        // allow empty string, in which case the version won't be renamed
        if (!val) return true;
        return semver.valid(semver.coerce(val)) ? true : 'Please specify a semantic version.';
      },
    },
    {
      // if the existing version being updated is already the main version,
      // we can't switch that so we skip this question
      type: isUpdate?.is_stable ? null : 'confirm',
      name: 'is_stable',
      message: 'Would you like to make this version the main version for this project?',
    },
    {
      type: 'confirm',
      name: 'is_beta',
      message: 'Should this version be in beta?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be hidden.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_public',
      message: 'Would you like to make this version public?',
    },
    {
      type: (prev, values) => {
        // if user wants this version to be the main version
        // it can't also be deprecated.
        return values.is_stable ? null : 'confirm';
      },
      name: 'is_deprecated',
      message: 'Would you like to deprecate this version?',
    },
  ];
}
