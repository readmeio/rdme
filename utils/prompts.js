exports.generatePrompts = versionList => [
  {
    type: 'select',
    name: 'option',
    message:
      "We couldn't find a version in ReadMe matching the version in your OAS file. Would you like to use an existing version or create a new one?",
    choices: [
      { message: 'Use existing', value: 'update' },
      { message: 'Create a new version', value: 'create' },
    ],
  },
  {
    type: 'select',
    name: 'versionSelection',
    message: 'Select your desired version',
    skip() {
      return this.enquirer.answers.option !== 'update';
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
];

exports.createVersionPrompt = (versionList, opts, isUpdate) => [
  {
    type: 'select',
    name: 'from',
    message: 'Which version would you like to fork from?',
    skip() {
      return opts.fork || isUpdate;
    },
    choices: versionList.map(v => {
      return {
        message: v.version,
        value: v.version,
      };
    }),
  },
  {
    type: 'input',
    name: 'newVersion',
    message: "What's your new version?",
    skip() {
      return opts.newVersion || !isUpdate;
    },
    hint: '1.0.0',
  },
  {
    type: 'confirm',
    name: 'is_stable',
    message: 'Would you like to make this version the main version for this project?',
    skip() {
      return opts.main || (isUpdate && isUpdate.is_stable);
    },
  },
  {
    type: 'confirm',
    name: 'is_beta',
    message: 'Should this version be in beta?',
    skip: () => opts.beta,
  },
  {
    type: 'confirm',
    name: 'is_hidden',
    message: 'Would you like to make this version public?',
    skip() {
      return opts.isPublic || opts.main || this.enquirer.answers.is_stable;
    },
  },
  {
    type: 'confirm',
    name: 'is_deprecated',
    message: 'Would you like to deprecate this version?',
    skip() {
      return opts.deprecated || opts.main || !isUpdate || this.enquirer.answers.is_stable;
    },
  },
];
