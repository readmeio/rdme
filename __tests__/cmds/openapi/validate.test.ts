/* eslint-disable no-console */
import fs from 'fs';

import chalk from 'chalk';
import prompts from 'prompts';

import OpenAPIValidateCommand from '../../../src/cmds/openapi/validate';
import ValidateAliasCommand from '../../../src/cmds/validate';
import { after, before } from '../../helpers/get-gha-setup';

const testWorkingDir = process.cwd();

const validate = new OpenAPIValidateCommand();
const validateAlias = new ValidateAliasCommand();

let consoleSpy;

const getCommandOutput = () => {
  return [consoleSpy.mock.calls.join('\n\n')].filter(Boolean).join('\n\n');
};

describe('rdme openapi:validate', () => {
  beforeEach(() => {
    consoleSpy = jest.spyOn(console, 'info').mockImplementation();
  });

  afterEach(() => {
    consoleSpy.mockRestore();

    process.chdir(testWorkingDir);
  });

  it.each([
    ['Swagger 2.0', 'json', '2.0'],
    ['Swagger 2.0', 'yaml', '2.0'],
    ['OpenAPI 3.0', 'json', '3.0'],
    ['OpenAPI 3.0', 'yaml', '3.0'],
    ['OpenAPI 3.1', 'json', '3.1'],
    ['OpenAPI 3.1', 'yaml', '3.1'],
  ])('should support validating a %s definition (format: %s)', (_, format, specVersion) => {
    expect(console.info).toHaveBeenCalledTimes(0);
    return expect(
      validate.run({
        spec: require.resolve(`@readme/oas-examples/${specVersion}/${format}/petstore.${format}`),
      })
    ).resolves.toContain(
      `petstore.${format} is a valid ${specVersion === '2.0' ? 'Swagger' : 'OpenAPI'} API definition!`
    );
  });

  it('should discover and upload an API definition if none is provided', async () => {
    await expect(validate.run({ workingDirectory: './__tests__/__fixtures__/relative-ref-oas' })).resolves.toBe(
      chalk.green('petstore.json is a valid OpenAPI API definition!')
    );

    expect(console.info).toHaveBeenCalledTimes(1);

    const output = getCommandOutput();
    expect(output).toBe(chalk.yellow('ℹ️  We found petstore.json and are attempting to validate it.'));
  });

  it('should select spec in prompt and validate it', async () => {
    const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
    prompts.inject([spec]);
    await expect(validate.run({})).resolves.toBe(chalk.green(`${spec} is a valid OpenAPI API definition!`));
  });

  it('should use specified working directory', () => {
    return expect(
      validate.run({
        spec: 'petstore.json',
        workingDirectory: './__tests__/__fixtures__/relative-ref-oas',
      })
    ).resolves.toBe(chalk.green('petstore.json is a valid OpenAPI API definition!'));
  });

  it('should adhere to .gitignore in subdirectories', () => {
    fs.copyFileSync(
      require.resolve('@readme/oas-examples/3.0/json/petstore-simple.json'),
      './__tests__/__fixtures__/nested-gitignored-oas/nest/petstore-ignored.json'
    );

    return expect(
      validate.run({
        workingDirectory: './__tests__/__fixtures__/nested-gitignored-oas',
      })
    ).resolves.toBe(chalk.green('nest/petstore.json is a valid OpenAPI API definition!'));
  });

  describe('error handling', () => {
    it('should throw an error if invalid JSON is supplied', () => {
      return expect(validate.run({ spec: './__tests__/__fixtures__/invalid-json/yikes.json' })).rejects.toStrictEqual(
        new SyntaxError('Unexpected end of JSON input')
      );
    });

    it('should throw an error if an invalid OpenAPI 3.0 definition is supplied', () => {
      return expect(validate.run({ spec: './__tests__/__fixtures__/invalid-oas.json' })).rejects.toThrow(
        'Token "Error" does not exist.'
      );
    });

    it('should throw an error if an invalid OpenAPI 3.1 definition is supplied', () => {
      return expect(validate.run({ spec: './__tests__/__fixtures__/invalid-oas-3.1.json' })).rejects.toMatchSnapshot();
    });

    it('should throw an error if an invalid Swagger definition is supplied', () => {
      return expect(validate.run({ spec: './__tests__/__fixtures__/invalid-swagger.json' })).rejects.toMatchSnapshot();
    });

    it('should throw an error if an invalid API definition has many errors', () => {
      return expect(validate.run({ spec: './__tests__/__fixtures__/very-invalid-oas.json' })).rejects.toMatchSnapshot();
    });
  });

  describe('CI tests', () => {
    beforeEach(() => {
      process.env.TEST_CI = 'true';
    });

    afterEach(() => {
      delete process.env.TEST_CI;
    });

    it('should successfully validate prompt and not run GHA onboarding', async () => {
      process.env.TEST_CREATEGHA = 'true';
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      await expect(validate.run({ spec })).resolves.toBe(chalk.green(`${spec} is a valid OpenAPI API definition!`));
      delete process.env.TEST_CREATEGHA;
    });

    it('should fail if user attempts to pass `--github` flag in CI environment', () => {
      return expect(
        validate.run({ github: true, spec: '__tests__/__fixtures__/petstore-simple-weird-version.json' })
      ).rejects.toStrictEqual(new Error('The `--github` flag is only for usage in non-CI environments.'));
    });
  });

  describe('GHA onboarding E2E tests', () => {
    let yamlOutput;

    beforeEach(() => {
      before((fileName, data) => {
        yamlOutput = data;
      });
    });

    afterEach(() => {
      after();
    });

    it('should create GHA workflow if user passes in spec via prompts', async () => {
      expect.assertions(6);
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      const fileName = 'validate-test-file';
      prompts.inject([spec, true, 'validate-test-branch', fileName]);

      await expect(validate.run({})).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
      expect(console.info).toHaveBeenCalledTimes(2);
      const output = getCommandOutput();
      expect(output).toMatch("Looks like you're running this command in a GitHub Repository!");
      expect(output).toMatch('is a valid OpenAPI API definition!');
    });

    it('should create GHA workflow if user passes in spec via opt', async () => {
      expect.assertions(3);
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      const fileName = 'validate-test-opt-spec-file';
      prompts.inject([true, 'validate-test-opt-spec-branch', fileName]);

      await expect(validate.run({ spec })).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
    });

    it('should create GHA workflow if user passes in spec via opt (github flag enabled)', async () => {
      expect.assertions(3);
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      const fileName = 'validate-test-opt-spec-github-file';
      prompts.inject(['validate-test-opt-spec-github-branch', fileName]);

      await expect(validate.run({ spec, github: true })).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
    });

    it('should create GHA workflow if user passes in spec via opt (including workingDirectory)', async () => {
      expect.assertions(3);
      const spec = 'petstore.json';
      const fileName = 'validate-test-opt-spec-workdir-file';
      prompts.inject([true, 'validate-test-opt-spec-github-branch', fileName]);

      await expect(
        validate.run({ spec, workingDirectory: './__tests__/__fixtures__/relative-ref-oas' })
      ).resolves.toMatchSnapshot();

      expect(yamlOutput).toMatchSnapshot();
      expect(fs.writeFileSync).toHaveBeenCalledWith(`.github/workflows/${fileName}.yml`, expect.any(String));
    });

    it('should reject if user says no to creating GHA workflow', () => {
      const spec = '__tests__/__fixtures__/petstore-simple-weird-version.json';
      prompts.inject([spec, false]);
      return expect(validate.run({})).rejects.toStrictEqual(
        new Error(
          'GitHub Actions workflow creation cancelled. If you ever change your mind, you can run this command again with the `--github` flag.'
        )
      );
    });
  });
});

describe('rdme validate', () => {
  let consoleWarnSpy;

  const getWarningCommandOutput = () => {
    return [consoleWarnSpy.mock.calls.join('\n\n')].filter(Boolean).join('\n\n');
  };

  beforeEach(() => {
    consoleWarnSpy = jest.spyOn(console, 'warn').mockImplementation();
  });

  afterEach(() => {
    consoleWarnSpy.mockRestore();
  });

  it('should should `rdme openapi:validate`', async () => {
    await expect(
      validateAlias.run({
        spec: require.resolve('@readme/oas-examples/3.0/json/petstore.json'),
      })
    ).resolves.toContain(chalk.green('petstore.json is a valid OpenAPI API definition!'));

    expect(console.warn).toHaveBeenCalledTimes(1);

    const output = getWarningCommandOutput();
    expect(output).toBe(
      chalk.yellow('⚠️  Warning! `rdme validate` has been deprecated. Please use `rdme openapi:validate` instead.')
    );
  });
});
