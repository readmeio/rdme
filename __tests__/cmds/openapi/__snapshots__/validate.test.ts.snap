// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`rdme openapi:validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt (github flag enabled) 1`] = `
"
Your GitHub Actions workflow file has been created! ✨

Almost done! Push your newly created file (.github/workflows/validate-test-opt-spec-github-file.yml) to GitHub and you're all set 🚀

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
"
`;

exports[`rdme openapi:validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt (github flag enabled) 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`validate-test-opt-spec-github-branch\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - validate-test-opt-spec-github-branch

jobs:
  rdme-openapi-validate:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v3

      - name: Run \`openapi:validate\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi:validate __tests__/__fixtures__/petstore-simple-weird-version.json
"
`;

exports[`rdme openapi:validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt 1`] = `
"
Your GitHub Actions workflow file has been created! ✨

Almost done! Push your newly created file (.github/workflows/validate-test-opt-spec-file.yml) to GitHub and you're all set 🚀

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
"
`;

exports[`rdme openapi:validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via opt 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`validate-test-opt-spec-branch\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - validate-test-opt-spec-branch

jobs:
  rdme-openapi-validate:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v3

      - name: Run \`openapi:validate\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi:validate __tests__/__fixtures__/petstore-simple-weird-version.json
"
`;

exports[`rdme openapi:validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via prompts 1`] = `
"
Your GitHub Actions workflow file has been created! ✨

Almost done! Push your newly created file (.github/workflows/validate-test-file.yml) to GitHub and you're all set 🚀

🦉 If you have any more questions, feel free to drop us a line! support@readme.io
"
`;

exports[`rdme openapi:validate > GHA onboarding E2E tests > should create GHA workflow if user passes in spec via prompts 2`] = `
"# This GitHub Actions workflow was auto-generated by the \`rdme\` cli on 2022-01-01T00:00:00.000Z
# You can view our full documentation here: https://docs.readme.com/docs/rdme
name: ReadMe GitHub Action 🦉

on:
  push:
    branches:
      # This workflow will run every time you push code to the following branch: \`validate-test-branch\`
      # Check out GitHub's docs for more info on configuring this:
      # https://docs.github.com/actions/using-workflows/events-that-trigger-workflows
      - validate-test-branch

jobs:
  rdme-openapi-validate:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repo 📚
        uses: actions/checkout@v3

      - name: Run \`openapi:validate\` command 🚀
        uses: readmeio/rdme@v7
        with:
          rdme: openapi:validate __tests__/__fixtures__/petstore-simple-weird-version.json
"
`;

exports[`rdme openapi:validate > error handling > should throw an error if an invalid API definition has many errors 1`] = `
[SyntaxError: OpenAPI schema validation failed.

REQUIRED must have required property 'url'

   7 |   },
   8 |   "servers": [
>  9 |     {
     |     ^ ☹️  url is missing here!
  10 |       "urll": "http://petstore.swagger.io/v2"
  11 |     }
  12 |   ],

ADDITIONAL PROPERTY must NOT have additional properties

  24 |     "/pet/findByStatus": {
  25 |       "get": {
> 26 |         "tagss": [
     |         ^^^^^^^ 😲  tagss is not expected to be here!
  27 |           "pet"
  28 |         ],
  29 |         "summary": "Finds Pets by status",]
`;

exports[`rdme openapi:validate > error handling > should throw an error if an invalid OpenAPI 3.1 definition is supplied 1`] = `
[SyntaxError: OpenAPI schema validation failed.

REQUIRED must have required property 'name'

  24 |   "components": {
  25 |     "securitySchemes": {
> 26 |       "tlsAuth": {
     |                  ^ ☹️  name is missing here!
  27 |         "type": "mutualTLS"
  28 |       }
  29 |     }]
`;

exports[`rdme openapi:validate > error handling > should throw an error if an invalid Swagger definition is supplied 1`] = `
[SyntaxError: Swagger schema validation failed.

ADDITIONAL PROPERTY must NOT have additional properties

  23 |         "description": "Find out more",
  24 |         "url": "http://swagger.io",
> 25 |         "this-shouldnt-be-here": true
     |         ^^^^^^^^^^^^^^^^^^^^^^^ 😲  this-shouldnt-be-here is not expected to be here!
  26 |       }
  27 |     }
  28 |   ],]
`;
